File: battleship/cpp/Board.cpp
#include <iostream> // for debugging
#include <stdexcept>
#include <utility> // for pair
#include <vector>
#include <random>
//#include <ctime>
//#include <unistd.h>
#include <string> // for std::to_string()

#include "Board.hpp"


std::pair<int, std::string> Board::checkHitUpdate(const std::pair<int, int>& coord){
  int hitCode = -1;
  std::string shipName = "";
  for(int i = 0; i < m_ships.size(); ++i){
    hitCode = m_ships[i]->checkHitUpdate(coord);
    if(hitCode != -1){
      // then it's a hit
      shipName = m_ships[i]->getName();
      if(hitCode == 1){
	// then it's a kill
	m_ships.erase(m_ships.begin() + i);
	if(m_ships.size() == 0){
	  // the the player lost
	  m_alive = false;
	}
      }
      break;
    }
  }
  
  return std::pair<int, std::string>(hitCode, shipName);
}


int Board::getShots() const{
  // Returns the number of shots available to the player.
  //  std::cout << "getShots()" << std::endl;
  int shots = 0;
  for(Ship* s : m_ships){
    shots += s->getShots();
  }
  return shots;
}


void Board::init(const int board_width, const int board_height){
  //  std::cout << "init(board_width, board_height)" << std::endl;
  m_alive = true;
  
  if(board_width < 1 || board_height < 1){
    throw std::invalid_argument("specified board dimensions are out of range...");
  }
  m_board_width = board_width;
  m_board_height = board_height;
}


void Board::init(const int board_width, const int board_height, const std::vector<Ship*>& ships){
  //  std::cout << "init(board_width, board_height, ships)" << std::endl;
  Board::init(board_width, board_height);
  Board::init(ships);
}


void Board::init(const std::vector<Ship*>& ships){
  //  std::cout << "init(ships)" << std::endl;
  m_ships = ships;
  Board::layoutShips();
}


void Board::layoutShips(){
  // layout ships on the board
  //
  // go through the list of ships, choose a random location and orientation, try to place the ship.
  // if the ship cells contain a cell of another ship already on the board, try again until timeout
  //  std::cout << "layoutShips()" << std::endl;
  std::vector<std::pair<int, int> > cells;
  for(Ship* s : m_ships){
    if(s != nullptr){
      for(int i = 0; i < SHIP_LAYOUT_TIMEOUT; ++i){
	int size = s->getSize();
	cells = Board::makeCellsVec(size);
	if(!Board::checkShipOverlap(cells)){
	  s->setCells(cells);
	  break;
	}
      }
    }else{
      // ship is nullptr
      std::cout << "ship is nullptr while laying out ships..." << std::endl;
    }
  }
}

std::vector<std::pair<int, int> > Board::makeCellsVec(const int size){
  // Returns a vector of pairs representing coordinates of cells of a ship.
  //
  // The cells are chosen at a random origin on the board and layed out
  // either vertically or horizontally.
  //  std::cout << "makeCellsVec(size)" << std::endl;
  std::vector<std::pair<int, int> > cells;
  if(size > m_board_width && size > m_board_height){
    throw std::invalid_argument("can't place ship... specified ship size is greater than the board size.\nboard_width = " + std::to_string(m_board_width) + "\tboard_height = " + std::to_string(m_board_height) + "\tship_size = " + std::to_string(size));
  }

  int orientVertical = rand() % 2;
  if(orientVertical){
    if(size > m_board_height){
      orientVertical = false;
    }    
  }else{
    if(size > m_board_width){
      orientVertical = true;
    }
  }
  std::pair<int, int> origin;
  if(orientVertical){
    //    std::cout << "orienting vertical" << std::endl;
    //    std::cout << "board_height - (size - 1) = " << std::to_string(board_height - (size - 1)) << std::endl;
    origin = std::pair<int, int>(rand() % m_board_width, rand() % (m_board_height - (size - 1)));
    //    std::cout << "origin x: " << origin.first << "\ty: " << origin.second << std::endl;
    cells.push_back(origin);
    for(int i = 1; i < size; ++i){
      cells.push_back(std::pair<int, int>(origin.first, origin.second + i));
    }
  }else{
    //    std::cout << "orienting horizontal" << std::endl;
    //    std::cout << "board_width - (size - 1) = " << std::to_string(board_height - (size - 1)) << std::endl;
    origin = std::pair<int, int>(rand() % (m_board_width - (size - 1)), rand() % m_board_height);
    //    std::cout << "origin x: " << origin.first << "\ty: " << origin.second << std::endl;
    cells.push_back(origin);
    for(int i = 1; i < size; ++i){
      cells.push_back(std::pair<int, int>(origin.first + i, origin.second));
    }
  }
  //  std::cout << "exiting makeCellsVec()" << std::endl;
  return cells;
}

bool Board::checkShipOverlap(const std::vector<std::pair<int, int> > cells){
  // Returns true if any coordinate in cells overlaps with a coordinate owned by another ship.
  // False otherwise.
  //  std::cout << "checkShipOverlap(cells)" << std::endl;
  std::vector<std::pair<int, int> > temp_cells;
  std::pair<int, int> temp_coord;
  for(Ship* s : m_ships){
    temp_cells = s->getCells();
    for(std::pair<int, int> temp_coord : temp_cells){
      for(std::pair<int, int> coord : cells){
	if(coord == temp_coord){
	  return true;
	}
      }
    }
  }
  return false;
}

File: battleship/java/Board.java
import java.util.ArrayList;

class Board{
    
    private String boardName;
    private int boardWidth;
    private int boardHeight;
    private ArrayList<Ship> ships;
    private static int MAX_BOARD_WIDTH = 100;
    private static int MAX_BOARD_HEIGHT = 100;
    private boolean boardAlive;
    private static int SHIP_PLACING_TIMEOUT = 1000; // number of times to try placing a ship
    
    public Board(int w, int h, String name) throws IllegalArgumentException{
	init(w, h, name);
    }
    
    public static void main(String[] args){
	testBoardClass(args);
    }
        
    public boolean addShip(Ship s){
	//
	// Places a ship on the board.
	// Returns true if the ship was able to be placed, false otherwise.
	//
	if(s.getSize() > this.boardWidth || s.getSize() > this.boardHeight){
	    // throw exception
	}
	ArrayList<Pair<Integer, Integer>> temp_cells;
	for(int i = 0; i < SHIP_PLACING_TIMEOUT; ++i){
	    Pair<Integer, Integer> pos = new Pair<Integer, Integer>( (int) (Math.random() * this.boardWidth), (int) (Math.random() * this.boardHeight) );
	    int dir = (int) (Math.random() * 2);
	    temp_cells = makeCellsList(pos, dir, s.getSize(), this.boardWidth, this.boardHeight);
	    if(temp_cells != null && !checkConflicting(temp_cells)){
		s.setCells(temp_cells);
		this.ships.add(s);
		return true;
	    }
	}
	return false;
    }
    
    private static ArrayList<Pair<Integer, Integer>> makeCellsList(Pair<Integer, Integer> pos, int dir, int size, int width, int height){
	//
	// Makes a list of cells starting from pos, in direction dir, of length size.
	//
	// If dir == 0 the direction is interpreted as horizontal.
	// Otherwise direction is interpreted as vertical.
	//
	// If the cells would go beyond the size of the board then null is returned.
	//
	if(dir == 0){
	    if(pos.first() + size > width){
		return null;
	    }
	}else{
	    if(pos.second() + size > height){
		return null;
	    }
	}
	
	ArrayList<Pair<Integer, Integer>> cells = new ArrayList<Pair<Integer, Integer>>();
	cells.add(pos);
	int x = pos.first();
	int y = pos.second();
	for(int i = 1; i < size; ++i){
	    if(dir == 0){
		x += 1;
	    }else{
		y += 1;
	    }
	    cells.add(new Pair<Integer, Integer>(x, y));
	}
	return cells;
    }
    
    private boolean checkConflicting(ArrayList<Pair<Integer, Integer>> cells){
	//
	// Checks if a ship overlaps with another ship already on the board.
	// Returns true if the ship overlaps, false otherwise.
	//
	for(Pair<Integer, Integer> pos : cells){
	    for(Ship s : this.ships){
		if(s.getCells().contains(pos)){
		    return true;
		}
	    }
	}
	return false;
    }
    
    public Pair<Integer, Ship> checkHit(Pair<Integer, Integer> pos){
	//
	// Checks if a position represents a hit on any ship on the board.
	//
	// Output: Pair<Integer, Ship>
	// If position is:
	//                 miss - Pair< -1, null>
	//                  hit - Pair<  0, Ship>
	//                 sink - Pair<  1, Ship>
	//
	for(Ship s : this.ships){
	    if(s.checkHit(pos)){
		if(!s.isAlive()){
		    this.ships.remove(s);
		    if(this.ships.size() == 0){
			this.boardAlive = false;
		    }
		    return new Pair<Integer, Ship>(1, s);
		}
		return new Pair<Integer, Ship>(0, s);
	    }
	}
	return new Pair<Integer, Ship>(-1, null);
    }
    
    public String getName(){
	return this.boardName;
    }
    
    public int getHeight(){
	return this.boardHeight;
    }
    
    public int getMaxHeight(){
	return this.MAX_BOARD_HEIGHT;
    }
    
    public int getMaxWidth(){
	return this.MAX_BOARD_WIDTH;
    }
    
    public ArrayList<Ship> getShips(){
	return this.ships;
    }
    
    public int getShots(){
	int shots = 0;
	for(Ship s : this.ships){
	    shots += s.getShots();
	}
	return shots;
    }
    
    public int getWidth(){
	return this.boardWidth;
    }
    
    protected void init(int w, int h, String name) throws IllegalArgumentException{
	if(w > MAX_BOARD_WIDTH || w < 1){
	    throw new IllegalArgumentException("specified board width is out of range... 0 < w < " + MAX_BOARD_WIDTH + " but w = " + w);
	}
	if(h > MAX_BOARD_HEIGHT || h < 1){
	    throw new IllegalArgumentException("specified board height is out of range... 0 < h < " + MAX_BOARD_HEIGHT + " but h = " + h);
	}
	this.boardWidth = w;
	this.boardHeight = h;
	this.boardName = name;
	this.ships = new ArrayList<Ship>();
	this.boardAlive = true;	
    }
    
    public boolean isAlive(){
	return this.boardAlive;
    }
    
    public void setName(String name){
	this.boardName = name;
    }
    
    private static void testBoardClass(String[] args){
	// just to test the class
	if(args.length != 2){
	    System.out.println("usage: ./Board width height");
	    return;
	}
	int x;
	int y;
	try{
	    x = Integer.parseInt(args[0]);
	    y = Integer.parseInt(args[1]);
	}catch(NumberFormatException e){
	    System.out.println("usage: ./Board (int)width (int)height");
	    return;
	}
	Board b = new Board(x, y, "TEST");
	b.addShip(new Carrier());
	b.addShip(new Cruiser());
	b.addShip(new Destroyer());
	b.addShip(new Submarine());
	b.addShip(new Minesweeper());
	System.out.println(b.getShips());
	ArrayList<Ship> temp_ships = b.getShips();
	for(Ship s : temp_ships){
	    ArrayList<Pair<Integer, Integer>> pairs = s.getCells();
	    for(Pair<Integer, Integer> p : pairs){
		System.out.println("(" + p.first() + ", " + p.second() + ")");
	    }
	}	
    }
}
